libafl/src/mutators/string.rs

```diff
@@ -0,0 +1,533 @@
+pub mod unicode_categories {
+    #![allow(unused)]
+    #![allow(missing_docs)]
+
+    include!(concat!(env!("OUT_DIR"), "/unicode_categories.rs"));
+}
+
+/// Mutator which retains the general category of a randomly selected range of bytes
+#[derive(Debug, Default)]
+pub struct StringCategoryPreservingMutator;
+
+impl Named for StringCategoryPreservingMutator {
+    fn name(&self) -> &str {
+        "string-category-preserving"
+    }
+}
+
+const MAX_CHARS: usize = 16;
+
+fn choose_index<R: Rand>(rand: &mut R, bytes: &[u8]) -> Option<(char, usize)> {
+    if bytes.is_empty() {
+        None
+    } else {
+        let idx = rand.below(bytes.len() as u64);
+        let mut indices = (idx.saturating_sub(3)..=idx).rev();
+
+        // try to find a character between [idx-3, idx+4)
+        let (c, idx, max_idx) = 'outerloop: loop {
+            if let Some(idx) = indices.next() {
+                let idx = idx as usize;
+                for max_idx in (idx + 1)..=(idx + 3).min(bytes.len()) {
+                    if let Ok(string) = core::str::from_utf8(&bytes[idx..max_idx]) {
+                        // we found a valid character
+                        break 'outerloop (string.chars().next().unwrap(), idx, max_idx);
+                    }
+                }
+            } else {
+                return None;
+            }
+        };
+        debug_assert!(idx + c.len_utf8() == max_idx);
+
+        Some((c, idx))
+    }
+}
+
+fn get_subcategory<T: Ord + Copy>(needle: T, haystack: &[(T, T)]) -> Option<(T, T)> {
+    haystack
+        .binary_search_by(|&(min, max)| match min.cmp(&needle) {
+            Ordering::Less | Ordering::Equal => match needle.cmp(&max) {
+                Ordering::Less | Ordering::Equal => Ordering::Equal,
+                Ordering::Greater => Ordering::Less,
+            },
+            Ordering::Greater => Ordering::Greater,
+        })
+        .ok()
+        .map(|idx| haystack[idx])
+}
+
+fn find_range<F: Fn(char) -> bool>(
+    bytes: &[u8],
+    c: char,
+    idx: usize,
+    predicate: F,
+) -> Range<usize> {
+    // walk backwards and discover
+    let max_idx = idx + c.len_utf8();
+    let mut backtracking = 0;
+    let mut curr_start = idx;
+    let mut curr_max = max_idx;
+    let mut start_iter = (0..idx).rev();
+    let start = loop {
+        if let Some(maybe_start) = start_iter.next() {
+            if let Ok(string) = core::str::from_utf8(&bytes[maybe_start..curr_max]) {
+                if let Some(first) = string.char_indices().next() {
+                    if let Some(last) = string.char_indices().next() {
+                        if curr_max == last.0 + last.1.len_utf8() && c == last.1 {
+                            if predicate(first.1) {
+                                curr_start = first.0;
+                                curr_max = last.0;
+                            }
+                        }
+                    }
+                }
+            }
+            backtracking += 1;
```

Comment:
does this mean we'll only go back for 4 bytes?

Reply:
That's correct; a valid code point will be at most 4 bytes.

Reply:
So in the end we're looking at the range that contains 3 UTF-8 Chars at maximum?
The character `c`, the char before `c` (we find it with backtracking) and then the char after `c`

Reply:
No, this selects a _single_ code point which intersects with the random byte we picked. Then we search for the range of characters around it of arbitrary length.

Reply:
I no longer use this algorithm as it is quite faulty.